Remove and replace most network topologies

AndrejMitrovic opened the pull request:
This will enable writing a few tests to finally close issue #716.

AndrejMitrovic committed:
Remove and replace most network topologies

This decouples creating nodes and setting
their initial known network state.

It enables building arbitrary amount of
outsider validator nodes, which is a
prerequisite for adding quorum tests
with many validator nodes.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -974,108 +972,61 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
-            .map!(idx => makeConfig(node_configs[idx], node_configs)).array;
-        break;
-    }
+    test_conf.validators.iota
+        .each!(_ => node_configs ~= makeNodeConfig(true, nextKey()));
+    test_conf.full_nodes.iota
+        .each!(_ => node_configs ~= makeNodeConfig(false, nextKey()));
```
That looks like a prime use case for `zip`.

Geod24 replied:
Or `refRange`

AndrejMitrovic replied:
But refrange doesn't change the semantics..

AndrejMitrovic replied:
Or I just don't know how to use it properly?

```D
import std.range;
import std.stdio;

void main ()
{
    auto keys = iota(5);

    auto r = refRange(&keys);

    foreach (x; r)
        writeln(x);
    foreach (x; r)
        writeln(x);
}
```

Prints the same range twice.

Geod24 replied:
```D
auto keys = WK.Keys.byRange().refRange();
keys.take(test_conf.validators).each!(key => node_configs ~= makeNodeConfig(true, key));
keys.take(test_conf.full_nodes).each!(key => node_configs ~= makeNodeConfig(false, key));
```

Would this work ?

AndrejMitrovic replied:
First of all refRange takes a pointer.

AndrejMitrovic replied:
(don't shoot the messenger)

Geod24 replied:
You got bamboozled by `foreach`. I have a bug report open for this.
```D
import std.range;
import std.stdio;
import std.algorithm;

void main ()
{
    auto keys = iota(5);

    auto r = refRange(&keys);
    r.each!writeln;
    r.each!writeln;
}
```

FTFY

AndrejMitrovic replied:
> /Users/andrejmitrovic/Downloads/ldc2-1.22.0-osx-x86_64/bin/../import/std/range/package.d(11502,13): Error: cannot modify struct instance *this._range of type Range because it contains const or immutable members

haha ranges go compiler errrrrror

AndrejMitrovic replied:
I'm reducing it.

AndrejMitrovic replied:
Reduced test-case:

```D
import std.range;

public static auto byRange ()
{
    static struct Range
    {
        size_t lbound = 0;

        // trigger point (remove const and it compiles)
        const size_t hbound = 100;

        size_t length () const { return 0; }
        bool empty () const { return false; }
        int front () const { return 0; }
        void popFront () { }
    }

    return Range(0);
}

void main ()
{
    auto keys = byRange();
    auto r = refRange(&keys);
}
```

Time to submit it..

AndrejMitrovic replied:
we can also work around it by making `hbound` an enum

Geod24 commented on source/agora/test/NetworkManager.d:
```diff
@@ -180,6 +180,7 @@ unittest
     // as the blocks in node_bad do not pass validation
-    retryFor(node_test.getBlockHeight() == 10, 4.seconds);
+    retryFor(node_test.getBlockHeight() == 10, 4.seconds,
+        node_test.getBlockHeight().to!string);
```
Yo

AndrejMitrovic replied:
Sorry. But it was useful during testing. :P

Geod24 reviewed (approved):
Sweeeeeeet, love it!

![image](https://user-images.githubusercontent.com/2180215/88627276-06e64b00-d0e7-11ea-8536-c4e2a5114157.png)

linked0 commented on source/agora/test/Base.d:
```diff
@@ -843,42 +843,36 @@ public class TestValidatorNode : Validator, TestAPI
+    /// and additionally outsider nodes will connect to them.
+    FullyConnected,
+
+    /// The nodes are connected in a chain: v1 < v2 < v3 < v1,
```
The `<` looks like an inequality sign, so how about using `<-`.

linked0 commented on source/agora/test/Base.d:
```diff
@@ -843,42 +843,36 @@ public class TestValidatorNode : Validator, TestAPI
+
+    /// The nodes are connected in a chain: v1 < v2 < v3 < v1,
+    /// and additionally outsider nodes are minimally connected to them
+    /// via v1 < o1, v2 < o2 (o = outsider node)
```
Same as the above comment.

linked0 commented on source/agora/test/Base.d:
```diff
@@ -974,108 +972,61 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
+    auto keys = WK.Keys.byRange();
+    auto nextKey () { scope (exit) keys.popFront(); return keys.front; }
     NodeConfig[] node_configs;
     Config[] configs;
```
I know the name `configs` has been used from start but it looks confusing. The `node_configs` and `configs` were confusing to me as a newcomer. How about using `main_configs` instead of `configs`.

AndrejMitrovic replied:
I actually had this change in mind too! I'll fix it.

linked0 reviewed (commented):
Great!
And, I have some comments.

AndrejMitrovic committed:
Change hbound to be an enum

Work around bug found in https://github.com/bpfkorea/agora/pull/1070#discussion_r461357617

AndrejMitrovic committed:
Rename config => main_config in tests

It makes it more obvious that they
are related to node configs.

AndrejMitrovic commented:
Fixed `const` => `enum` for the workaround, using `refRange` now, fixed the `<` to `<-`, and renamed `config` => `main_config`.
